WalletWasabi/Models/Label.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace WalletWasabi.Models
+{
+	public class Label : IEquatable<Label>
+	{
+		public static Label Empty { get; } = new Label();
+		public static char[] Separators { get; } = new[] { ',', ':' };
+		public IEnumerable<string> Labels { get; }
+		public bool IsEmpty { get; }
+
+		private string LabelString { get; }
+
+		public Label(params string[] labels) : this(labels as IEnumerable<string>)
+		{
+		}
+
+		public Label(IEnumerable<string> labels)
+		{
+			Labels = (labels
+				   ?.SelectMany(x => x?.Split(Separators, StringSplitOptions.RemoveEmptyEntries) ?? new string[0])
+				   ?.Select(x => x.Trim())
+				   ?.Where(x => x != "")
+				   ?.Distinct(StringComparer.OrdinalIgnoreCase)
+				   ?.OrderBy(x => x)
+				   ?? Enumerable.Empty<string>())
+				   .ToArray();
+
+			HashCode = ((IStructuralEquatable)Labels).GetHashCode(EqualityComparer<string>.Default);
+			IsEmpty = !Labels.Any();
+
+			LabelString = string.Join(", ", Labels);
+		}
+
+		public override string ToString() => LabelString;
+
+		public static Label Merge(IEnumerable<Label> labels)
+		{
+			return new Label(labels?
+				.SelectMany(x => x?.Labels ?? Enumerable.Empty<string>())
+				.Where(x => x != null)
+				?? Enumerable.Empty<string>());
+		}
+
+		public static Label Merge(params Label[] labels) => Merge(labels as IEnumerable<Label>);
+
+		#region Equality
+
+		public override bool Equals(object obj) => obj is Label label && this == label;
+
+		public bool Equals(Label other) => this == other;
+
+		private int HashCode { get; }
```

Comment:
Why not use a private field `private int _hashCode;`?
Also maybe it should be moved abode the constructor.

Reply:
> Why not use a private field private int _hashCode;?

Consistency with the rest of the codebase.

> Also maybe it should be moved abode the constructor.

Yeah, maybe, I would prefer to keep it in `#region Equality`, because it's almost always a a computational property (`public override int GetHashCode() =>`), but this code has to be as performant as possible and a design principle of this class is to follow the immutability of strings, so it's also "immutable", so we are able to only compute the `HashCode` in the constructor and don't bother about changes, because there won't be.